zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/SpotRestorationTool.java

```diff
@@ -0,0 +1,276 @@
+ * snapshot and transaction logs.
+ */
+public class SpotRestorationTool {
+  private static final Logger LOG = LoggerFactory.getLogger(SpotRestorationTool.class);
+
+  File dataDir;
+  ZooKeeper zk;
+  FileTxnSnapLog snapLog;
+  boolean restoreRecursively;
+  String targetZNodePath;
+  List<String> messages;
+
+  /**
+   * Constructor
+   * @param dataDir The directory contains snapshot and transaction log files to be used for restoration
+   * @param zk A client connection to the zk server to be restored
+   * @param targetZNodePath The znode path to start the restoration
+   * @param restoreRecursively If true then restore the whole sub-data tree of the target znode;
+   *                           if false then restore the target znode only
+   * @throws IOException
+   */
+  public SpotRestorationTool(File dataDir, ZooKeeper zk, String targetZNodePath,
+      boolean restoreRecursively) throws IOException {
+    if (dataDir == null || !dataDir.exists()) {
+      throw new IllegalArgumentException(
+          "The provided dataDir is null or does not exist, please check the input.");
+    }
+    this.dataDir = dataDir;
+    this.zk = zk;
+    this.snapLog = new FileTxnSnapLog(dataDir, dataDir);
+    this.targetZNodePath = targetZNodePath;
+    this.restoreRecursively = restoreRecursively;
+    this.messages = Lists.newArrayList();
+  }
+
+  /**
+   * Run the spot restoration.
+   * 1. If a node exists in restored data tree but not in the zk server, create the node in the server
+   * 2. If a node exists in zk server but not the restored data tree, show messages ask user to manually delete the node
+   * 3. If a node exists in both zk server and restored data tree, show message ask user to take care of the node
+   * @throws IOException
+   */
+  public void run() throws IOException {
+    DataTree dataTree = new DataTree();
+    long zxidRestored =
+        snapLog.restore(dataTree, new ConcurrentHashMap<>(), new FileTxnSnapLog.PlayBackListener() {
+          @Override
+          public void onTxnLoaded(TxnHeader hdr, Record rec, TxnDigest digest) {
+            // Do nothing since we are trying to build a data tree in memory, not in actual zk server
+          }
+        });
+
+    DataNode originalNode = dataTree.getNode(targetZNodePath);
+    if (originalNode == null) {
+      return;
+    }
+    Stat originalNodeStat = new Stat();
+    originalNode.copyStat(originalNodeStat);
+
+    LOG.info(
+        "The restored zk data tree is constructed. The highest zxid restored is: " + zxidRestored);
+    LOG.info("Requested path node data: " + Arrays.toString(originalNode.getData()));
+    LOG.info(
+        "Requested path node children: " + Arrays.toString(originalNode.getChildren().toArray()));
+    LOG.info("Requested path node stat: " + originalNodeStat.toString());
+    String requestMsg = "Do you want to restore to this point? Enter \"yes\" or \"no\".";
+    String yesMsg = "Restoring ZooKeeper server...";
```

Comment:
Nit: "Restoring ZooKeeper server..." is a bit vague - perhaps we could say something like "Performing spot restoration of the ZNode: path: {}"...?